src/storage/TokensProvider.tsx

```diff
@@ -0,0 +1,275 @@
+      })
+    },
+    [currentAccount, restoredVisibleTokens, updateCSTokens],
+  )
+
+  const isActiveToken = useCallback(
+    (token: Token) =>
+      !!(
+        currentAccount &&
+        restoredVisibleTokens &&
+        restoredVisibleTokens[currentAccount.address]?.includes(getKey(token))
+      ),
+    [currentAccount, restoredVisibleTokens],
+  )
+
+  useAsync(async () => {
+    try {
+      const response = await restoreVisibleTokens()
+
+      setRestoredVisibleTokens(response)
+    } catch {}
+  }, [])
+
+  const updateWithSplTokens = useCallback(async () => {
+    if (!currentAccount?.address) return
+
+    try {
+      const response = await restoreTokensMetadata()
+
+      if (!response[currentAccount.address]) return []
+
+      setTokens((prevTokens) => {
+        const newTokens = [...prevTokens]
+
+        response[currentAccount.address].forEach((item) => {
+          const index = newTokens.findIndex(
+            (token) => token.type === item.symbol,
+          )
+
+          if (index === -1) {
+            newTokens.push({
+              type: item.symbol as Ticker,
+              balance: new Balance(
+                splTokensBalance[item.mintAddress],
+                new BaseCurrencyType(item.symbol as Ticker, 9),
+              ),
+              name: item.name,
+              staked: false,
+            })
+          }
+        })
+
+        return newTokens
+      })
+    } catch {}
+  }, [currentAccount, splTokensBalance])
+
+  useEffect(() => {
+    updateWithSplTokens()
+  }, [updateWithSplTokens])
+
+  const prevAccount = usePrevious(currentAccount?.address)
+  useEffect(() => {
+    if (prevAccount !== currentAccount?.address) {
+      setTokens(defaultTokens)
+    }
+  }, [currentAccount, defaultTokens, prevAccount])
+
+  const visibleTokens = useMemo(() => {
+    if (!currentAccount?.address || !restoredVisibleTokens) return []
+
+    if (!restoredVisibleTokens[currentAccount.address]) return []
+
+    return tokens.filter((token) => {
+      const key = getKey(token)
+
+      return restoredVisibleTokens[currentAccount.address].includes(key)
+    })
+  }, [currentAccount, restoredVisibleTokens, tokens])
+
+  useEffect(() => {
+    if (!currentAccount?.address || !restoredVisibleTokens) return
+
+    if (restoredVisibleTokens[currentAccount.address]) return
+
+    updateCSTokens({
+      ...restoredVisibleTokens,
+      [currentAccount.address]: tokens.map((token) => getKey(token)),
+    })
+  }, [currentAccount, restoredVisibleTokens, tokens, updateCSTokens])
+
+  const value = useMemo(
+    () => ({
+      tokens,
+      handleUpdateTokens,
+      visibleTokens,
+      isActiveToken,
+    }),
+    [tokens, handleUpdateTokens, visibleTokens, isActiveToken],
+  )
+
+  return value
+}
+
+const initialState = {
+  tokens: [],
+  handleUpdateTokens: () => {},
+  visibleTokens: [],
+  isActiveToken: () => false,
+}
+
+const TokensContext =
+  createContext<ReturnType<typeof useTokensHook>>(initialState)
```

Comment:
We are moving the app towards using `redux`. Do you think we could benefit off of making this a slice or adding this as part of an existing slice?